services/graph/pkg/service/v0/tags.go

```diff
@@ -0,0 +1,226 @@
+	revactx "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/cs3org/reva/v2/pkg/events"
+	"github.com/cs3org/reva/v2/pkg/storagespace"
+	revaTags "github.com/cs3org/reva/v2/pkg/tags"
+	"github.com/go-chi/render"
+	libregraph "github.com/owncloud/libre-graph-api-go"
+	searchsvc "github.com/owncloud/ocis/v2/protogen/gen/ocis/services/search/v0"
+	"github.com/owncloud/ocis/v2/services/graph/pkg/service/v0/errorcode"
+	"go-micro.dev/v4/metadata"
+)
+
+func (g Graph) GetTags(w http.ResponseWriter, r *http.Request) {
+	th := r.Header.Get(revactx.TokenHeader)
+	ctx := revactx.ContextSetToken(r.Context(), th)
+	ctx = metadata.Set(ctx, revactx.TokenHeader, th)
+	sr, err := g.searchService.Search(ctx, &searchsvc.SearchRequest{
+		Query:    "Tags:*",
+		PageSize: -1,
+	})
+	if err != nil {
+		g.logger.Error().Err(err).Msg("Could not search for tags")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	tagList := revaTags.FromList("")
+	for _, match := range sr.Matches {
+		for _, tag := range match.Entity.Tags {
+			tagList.AddList(tag)
+		}
+	}
+
+	tagCollection := libregraph.NewCollectionOfTags()
+	tagCollection.Value = tagList.AsSlice()
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, tagCollection)
+}
+
+func (g Graph) AssignTags(w http.ResponseWriter, r *http.Request) {
+	var (
+		assignment libregraph.TagAssignment
+		ctx        = r.Context()
+	)
+
+	if err := json.NewDecoder(r.Body).Decode(&assignment); err != nil {
+		g.logger.Debug().Err(err).Interface("body", r.Body).Msg("could not decode tag assignment request")
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "invalid body schema definition")
+		return
+	}
+
+	rid, err := storagespace.ParseID(assignment.ResourceId)
+	if err != nil {
+		g.logger.Debug().Err(err).Msg("could not parse resourceId")
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "invalid resourceId")
+		return
+	}
+
+	sres, err := g.gatewayClient.Stat(ctx, &provider.StatRequest{
+		Ref: &provider.Reference{ResourceId: &rid},
+	})
+	if err != nil {
+		g.logger.Error().Err(err).Msg("error stating file")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if sres.GetStatus().GetCode() != rpc.Code_CODE_OK {
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "can't stat resource")
+		return
+	}
+
+	pm := sres.GetInfo().GetPermissionSet()
+	if pm == nil {
+		g.logger.Error().Err(err).Msg("no permissionset on file")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// it says we need "write access" to set tags. One of those should do
+	if !pm.InitiateFileUpload && !pm.CreateContainer {
+		g.logger.Info().Msg("no permission to create a tag")
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	var currentTags string
+	if m := sres.GetInfo().GetArbitraryMetadata().GetMetadata(); m != nil {
+		currentTags = m["tags"]
+	}
+
+	allTags := revaTags.FromList(currentTags)
+	newTags := strings.Join(assignment.Tags, ",")
```

Comment:
Why is it needed to add a `,` to the tags string? Bug in `reva/tags` pkg?

Reply:
assignment.Tags is a slice of tags ([]string), tags package does not have a fromSlice func at the moment.

Reply:
Hehe. OK. So it IS a bug in `reva/tags` pkg :smile: